File: solutions/haskell/allergies/1/src/Allergies.hs
module Allergies (Allergen(..), allergies, isAllergicTo) where

data Allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats
              deriving (Eq, Show)

allergies :: Int -> [Allergen]
allergies score
 | score >= 2^7 = allergies (score - 2^7) ++ [Cats]
 | score >= 2^6 = allergies (score - 2^6) ++ [Pollen]
 | score >= 2^5 = allergies (score - 2^5) ++ [Chocolate]
 | score >= 2^4 = allergies (score - 2^4) ++ [Tomatoes]
 | score >= 2^3 = allergies (score - 2^3) ++ [Strawberries]
 | score >= 2^2 = allergies (score - 2^2) ++ [Shellfish]
 | score >= 2^1 = allergies (score - 2^1) ++ [Peanuts]
 | score >= 2^0 = [Eggs]
 | otherwise = []

isAllergicTo :: Allergen -> Int -> Bool
isAllergicTo allergen score = allergen `elem` allergies score

File: solutions/erlang/allergies/1/src/allergies.erl
-module(allergies).

-export([allergies/1, is_allergic_to/2]).

allergies(Score) when Score >= 512 -> allergies(Score rem 512);
allergies(Score) when Score >= 256 -> allergies(Score rem 256);
allergies(Score) when Score >= 128 -> allergies(Score rem 128) ++ [cats];
allergies(Score) when Score >= 64 -> allergies(Score - 64) ++ [pollen];
allergies(Score) when Score >= 32 -> allergies(Score - 32) ++ [chocolate];
allergies(Score) when Score >= 16 -> allergies(Score - 16) ++ [tomatoes];
allergies(Score) when Score >= 8 -> allergies(Score - 8) ++ [strawberries];
allergies(Score) when Score >= 4 -> allergies(Score - 4) ++ [shellfish];
allergies(Score) when Score >= 2 -> allergies(Score - 2) ++ [peanuts];
allergies(Score) when Score >= 1 -> [eggs];
allergies(_) -> [].

is_allergic_to(Substance, Score) -> lists:member(Substance, allergies(Score)).
